src/Hl7.Fhir.Support.Poco/Specification/Source/CachedResolver.cs

```diff
@@ -0,0 +1,307 @@
+        public bool InvalidateByUri(string url)
+        {
+            if (url == null) throw Error.ArgumentNull(nameof(url));
+            return _resourcesByUri.Invalidate(url);
+        }
+
+        /// <summary>Clear the cache entry for the conformance resource with the specified canonical uri, if it exists.</summary>
+        /// <param name="url">The canonical url of the target conformance resource.</param>
+        /// <returns><c>true</c> if succesful, <c>false</c> otherwise.</returns>
+        public bool InvalidateByCanonicalUri(string url)
+        {
+            if (url == null) throw Error.ArgumentNull(nameof(url));
+            return _resourcesByCanonical.Invalidate(url);
+        }
+
+        /// <summary>Clear the memory cache by removing all existing cache entries.</summary>
+        public void Clear()
+        {
+            _resourcesByUri.Clear();
+            _resourcesByCanonical.Clear();
+        }
+
+        /// <summary>Determines if the memory cache contains a resource with the specified url.</summary>
+        /// <returns><c>true</c> if the resource is cached, or <c>false</c> otherwise.</returns>
+        public bool IsCachedUri(string url) => _resourcesByUri.Contains(url);
+
+        /// <summary>Determines if the memory cache contains a conformance resource with the specified canonical url.</summary>
+        /// <returns><c>true</c> if the conformance resource is cached, or <c>false</c> otherwise.</returns>
+        public bool IsCachedCanonicalUri(string url) => _resourcesByCanonical.Contains(url);
+
+        /// <summary>Event arguments for the <see cref="LoadResourceEventHandler"/> delegate.</summary>
+        public class LoadResourceEventArgs : EventArgs
+        {
+            public LoadResourceEventArgs(string url, Resource resource) : base()
+            {
+                Url = url;
+                Resource = resource;
+            }
+
+            /// <summary>Returns the url of the cached resource.</summary>
+            public string Url { get; }
+
+            /// <summary>Returns a reference to the cached resource.</summary>
+            public Resource Resource { get; }
+        }
+
+        /// <summary>Handles the <see cref="Load"/> event that is fired when a new resources is loaded into the cache.</summary>
+        public delegate void LoadResourceEventHandler(object sender, LoadResourceEventArgs e);
+
+        /// <summary>Occurs when a artifact is loaded into the cache.</summary>
+        public event LoadResourceEventHandler Load;
+
+        /// <summary>Called when an artifact is loaded into the cache.</summary>
+        protected virtual void OnLoad(string url, Resource resource) => Load?.Invoke(this, new LoadResourceEventArgs(url, resource));
+
+        internal async Task<Resource> InternalResolveByUri(string url)
+        {
+            var resource = await AsyncResolver.ResolveByUriAsync(url);
+            OnLoad(url, resource);
```

Comment:
Should this also not be async?

Reply:
We should make this a private method - then we don't have to put the Async suffix.

Reply:
I meant the OnLoad. So `await OnLoad(url, resource);`. 

So, should we make OnLoad async?

Reply:
Note that the I/O has already been done, this is supposed to be just a quick signal to interested parties that I/O has finished, so no async necessary here.  (and I didn't want to go into async events yet).